File: LeetCode/515/Solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
private:
    vector<int> result;
public:
    vector<int> largestValues(TreeNode* root) {
        solve(root, 0);
        
        return result;
    }
    
    void solve(TreeNode* root, int depth) {
        if(root == NULL)
            return;
        
        if(result.size() == depth) {
            result.push_back(root->val);
        } else {
            if(result[depth] < root->val)
                result[depth] = root->val;
        }
        
        solve(root->left, depth + 1);
        solve(root->right, depth + 1);
    }
};

File: LeetCode/515/Solution.java
class Solution {
    public List<Integer> largestValues(TreeNode root) {
        if(root == null) {
            return new LinkedList<>();
        }

        Queue<TreeNode> q = new LinkedList<>();
        q.add(root);

        List<Integer> result = new ArrayList<>();
        while(!q.isEmpty()) {
            int size = q.size();
            int temp = Integer.MIN_VALUE;
            for(int i=0;i<size;++i) {
                TreeNode node = q.poll();
                temp = Math.max(temp, node.val);
                if(node.left != null) {
                    q.add(node.left);
                }
                if(node.right != null) {
                    q.add(node.right);
                }
            }
            result.add(temp);
        }
        return result;
    }
}
